src/simulation/dynamics/gravityEffector/sphericalHarmonicsGravityModel.cpp

```diff
@@ -0,0 +1,204 @@
+ */
+
+#include "sphericalHarmonicsGravityModel.h"
+#include "architecture/utilities/bskLogging.h"
+#include "simulation/dynamics/_GeneralModuleFiles/gravityEffector.h"
+
+namespace {
+// Computes the term (2 - d_l), where d_l is the kronecker delta.
+inline double getK(const unsigned int degree)
+{
+    return (degree == 0) ? 1.0 : 2.0;
+}
+}
+
+std::optional<std::string> SphericalHarmonicsGravityModel::initializeParameters()
+{
+    if (this->cBar.size() == 0 || this->sBar.size() == 0) {
+        return "Could not initialize spherical harmonics: the 'C' or 'S' parameters were not "
+               "provided.";
+    }
+
+    for (size_t i = 0; i <= this->maxDeg + 1; i++) {
+        std::vector<double> aRow, n1Row, n2Row;
+        aRow.resize(i + 1, 0.0);
+        // Diagonal elements of A_bar
+        if (i == 0) { aRow[i] = 1.0; }
+        else {
+            aRow[i] = sqrt(double((2 * i + 1) * getK(i)) / (2 * i * getK(i - 1))) *
+                      this->aBar[i - 1][i - 1];
+        }
+        n1Row.resize(i + 1, 0.0);
+        n2Row.resize(i + 1, 0.0);
+        for (size_t m = 0; m <= i; m++) {
+            if (i >= m + 2) {
+                n1Row[m] = sqrt(double((2 * i + 1) * (2 * i - 1)) / ((i - m) * (i + m)));
+                n2Row[m] = sqrt(double((i + m - 1) * (2 * i + 1) * (i - m - 1)) /
+                                ((i + m) * (i - m) * (2 * i - 3)));
+            }
+        }
+        this->n1.push_back(n1Row);
+        this->n2.push_back(n2Row);
+        this->aBar.push_back(aRow);
+    }
+
+    for (size_t l = 0; l <= this->maxDeg; l++) // up to _maxDegree-1
+    {
+        std::vector<double> nq1Row, nq2Row;
+        nq1Row.resize(l + 1, 0.0);
+        nq2Row.resize(l + 1, 0.0);
+        for (size_t m = 0; m <= l; m++) {
+            if (m < l) { nq1Row[m] = sqrt(double((l - m) * getK(m) * (l + m + 1)) / getK(m + 1)); }
+            nq2Row[m] = sqrt(double((l + m + 2) * (l + m + 1) * (2 * l + 1) * getK(m)) /
+                             ((2 * l + 3) * getK(m + 1)));
+        }
+        this->nQuot1.push_back(nq1Row);
+        this->nQuot2.push_back(nq2Row);
+    }
+
+    return {}; // No error!
+}
+
+std::optional<std::string>
+SphericalHarmonicsGravityModel::initializeParameters(const GravBodyData& body)
+{
+    this->radEquator = body.radEquator;
+    this->muBody = body.mu;
+    return this->initializeParameters();
+}
+
+Eigen::Vector3d
+SphericalHarmonicsGravityModel::computeField(const Eigen::Vector3d& position_planetFixed) const
+{
+    return this->computeField(position_planetFixed, this->maxDeg, true);
+}
+
+Eigen::Vector3d
+SphericalHarmonicsGravityModel::computeField(const Eigen::Vector3d& position_planetFixed,
+                                             unsigned int degree, bool include_zero_degree) const
+{
+    if (degree > this->maxDeg) {
+        auto errorMsg =
+            "Requested degree greater than maximum degree in Spherical Harmonics gravity model";
+        throw std::invalid_argument(errorMsg);
```

Comment:
We should be using BSK_ERROR error messages here?  Otherwise we have a mix of error handling in BSK?

Reply:
@patkenneally , just tagging you here as you said you see reasons to use `throw` at times.  Thanks for your input on this.

Reply:
Ok, spoke with @patkenneally today and he explained the benefits of using `throw`.  Let's leave what you have, but add BSK_ERROR logging as well.  That we have the ability to respond on a higher level to an error being thrown, but also have the logging functionality as other BSK modules use.